react/OrderItems.tsx

```diff
@@ -69,164 +76,308 @@ const maybeUpdateTotalizers = (
-  const { enqueue, listen, isWaiting } = useOrderQueue()
-  const { loading, orderForm, setOrderForm } = useOrderForm()
+const noop = async () => {}
 
-  const [updateItems] = useMutation(UpdateItems)
+const OrderItemsContext = createContext<Context>({
+  addItem: noop,
+  updateQuantity: noop,
+  removeItem: noop,
+})
 
+const useEnqueueTask = () => {
+  const { enqueue, listen } = useOrderQueue()
   const queueStatusRef = useQueueStatus(listen)
-  const lastUpdateTaskRef = useRef({
-    promise: undefined,
-    variables: undefined,
-  } as EnqueuedTask)
+  const { setOrderForm } = useOrderForm()
+
+  const enqueueTask = useCallback<(task: () => Promise<OrderForm>) => void>(
+    task => {
+      enqueue(task).then(
+        (orderForm: OrderForm) => {
+          popLocalOrderQueue()
+          if (queueStatusRef.current === QueueStatus.FULFILLED) {
+            setOrderForm(orderForm)
+          }
+        },
+        (error: any) => {
+          if (!error || error.code !== 'TASK_CANCELLED') {
+            throw error
+          }
+        }
+      )
+    },
+    [enqueue, queueStatusRef, setOrderForm]
+  )
 
-  if (!orderForm) {
-    throw new Error('Unable to fetch order form.')
-  }
+  return enqueueTask
+}
+
+interface UpdateItemsMutation {
+  updateItems: OrderForm
+}
 
-  const { items, totalizers, value: orderFormValue } = orderForm
+export const OrderItemsProvider: FC = ({ children }) => {
+  const { orderForm, setOrderForm } = useOrderForm()
 
-  const itemIds = useCallback(
-    (props: Partial<Item>) => {
-      let index = props.index
-      let uniqueId = props.uniqueId
+  const enqueueTask = useEnqueueTask()
 
-      if (index) {
-        uniqueId = items[index].uniqueId as string
-      } else if (uniqueId) {
-        index = items.findIndex(
-          (item: Item) => item.uniqueId === props.uniqueId
-        ) as number
-      } else {
-        throw new Error(
-          'Either index or uniqueId must be provided when updating an item'
+  const orderFormItemsRef = useRef(orderForm.items)
+
+  useEffect(() => {
+    orderFormItemsRef.current = orderForm.items
+  }, [orderForm.items])
+
+  const [mutateUpdateQuantity] = useMutation<UpdateItemsMutation>(UpdateItems)
+  const [mutateAddItem] = useMutation<
+    { addToCart: OrderForm },
+    { items: OrderFormItemInput[] }
+  >(AddToCart)
+
+  const addItem = useCallback(
+    (items: Partial<Item>[]) => {
+      const mutationInput = items.map(adjustForItemInput)
+
+      const orderFormItems = mutationInput
+        .map((itemInput, index) =>
+          mapItemInputToOrderFormItem(itemInput, items[index])
+        )
+        .filter(
+          orderFormItem =>
+            orderFormItemsRef.current.findIndex(
+              (item: any) => item.id === orderFormItem.id
+            ) === -1
         )
-      }
 
-      return { index, uniqueId }
-    },
-    [items]
-  )
+      if (orderFormItems.length === 0) {
+        // all items already exists in the minicart
```

Comment:
thinking about the possibility of having the same item in the cart twice, do you think I should remove this early return? I wouldn't know what to check to not add the item twice and preserve the same behaviour as the API though

Reply:
Can't say for sure because there might be a scenario I'm not aware of, but I think when _adding_ an item to the cart, if the item's id is already present, then it is always a noop, so leaving this early return should be fine.